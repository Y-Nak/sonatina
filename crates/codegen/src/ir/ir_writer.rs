use std::io;

use super::{Block, Function, Insn, InsnData, Type, Value};

pub struct FuncWriter<'a> {
    func: &'a Function,
    level: u8,
}

impl<'a> FuncWriter<'a> {
    pub fn new(func: &'a Function) -> Self {
        Self { func, level: 0 }
    }

    pub fn write(&mut self, mut w: impl io::Write) -> io::Result<()> {
        w.write_fmt(format_args!("func %{}(", self.func.name))?;
        let sig = &self.func.sig;
        self.write_iter_with_delim(sig.args().iter(), ", ", &mut w)?;

        let mut rets = sig.returns().iter().peekable();
        if rets.peek().is_some() {
            w.write_all(b") -> ")?;
        } else {
            w.write_all(b")")?;
        }

        self.write_iter_with_delim(rets, ", ", &mut w)?;

        self.enter_item(&mut w)?;
        for block in self.func.layout.iter_block() {
            self.write_block_with_insn(block, &mut w)?;
            self.newline(&mut w)?;
            self.newline(&mut w)?;
        }
        self.leave_item();

        Ok(())
    }

    pub fn dump_string(&mut self) -> io::Result<String> {
        let mut s = Vec::new();
        self.write(&mut s)?;
        unsafe { Ok(String::from_utf8_unchecked(s)) }
    }

    fn write_block_with_insn(&mut self, block: Block, mut w: impl io::Write) -> io::Result<()> {
        self.indent(&mut w)?;
        block.write(self, &mut w)?;

        self.enter_item(&mut w)?;
        let insns = self.func.layout.iter_insn(block);
        self.write_iter_with_delim(insns, "\n", &mut w)?;
        self.leave_item();

        Ok(())
    }

    fn write_insn_args(&mut self, args: &[Value], mut w: impl io::Write) -> io::Result<()> {
        self.write_iter_with_delim(args.iter(), " ", &mut w)
    }

    fn write_iter_with_delim<T>(
        &mut self,
        iter: impl Iterator<Item = T>,
        delim: &str,
        mut w: impl io::Write,
    ) -> io::Result<()>
    where
        T: IrWrite,
    {
        let mut iter = iter.peekable();
        while let Some(item) = iter.next() {
            item.write(self, &mut w)?;
            if iter.peek().is_some() {
                w.write_all(delim.as_bytes())?;
            }
        }

        Ok(())
    }

    fn indent(&self, mut w: impl io::Write) -> io::Result<()> {
        w.write_all(" ".repeat(self.level as usize * 4).as_bytes())
    }

    fn newline(&self, mut w: impl io::Write) -> io::Result<()> {
        w.write_all(b"\n")
    }

    fn space(&self, mut w: impl io::Write) -> io::Result<()> {
        w.write_all(b" ")
    }

    fn enter_item(&mut self, mut w: impl io::Write) -> io::Result<()> {
        self.level += 1;
        w.write_all(b":\n")
    }

    fn leave_item(&mut self) {
        self.level -= 1;
    }
}

trait IrWrite {
    fn write(&self, writer: &mut FuncWriter, w: impl io::Write) -> io::Result<()>;
}

impl IrWrite for Value {
    fn write(&self, writer: &mut FuncWriter, mut w: impl io::Write) -> io::Result<()> {
        w.write_fmt(format_args!(
            "v{}.{}",
            self.0,
            writer.func.dfg.value_ty(*self),
        ))
    }
}

impl IrWrite for Block {
    fn write(&self, _: &mut FuncWriter, mut w: impl io::Write) -> io::Result<()> {
        w.write_fmt(format_args!("block{}", self.0))
    }
}

impl IrWrite for Type {
    fn write(&self, _: &mut FuncWriter, mut w: impl io::Write) -> io::Result<()> {
        write!(w, "{}", self)
    }
}

impl IrWrite for Insn {
    fn write(&self, writer: &mut FuncWriter, mut w: impl io::Write) -> io::Result<()> {
        use InsnData::*;

        writer.indent(&mut w)?;
        let ret_val = writer.func.dfg.insn_result(*self);
        if let Some(ret_val) = ret_val {
            ret_val.write(writer, &mut w)?;
            w.write_all(b" = ")?;
        }

        let insn_data = writer.func.dfg.insn_data(*self);
        match insn_data {
            Immediate { code } => w.write_all(code.to_string().as_bytes())?,
            Binary { code, args } => {
                w.write_fmt(format_args!("{}", code.as_str()))?;
                writer.space(&mut w)?;
                writer.write_insn_args(args, &mut w)?;
            }
            Cast { code, args, .. } => {
                w.write_fmt(format_args!("{}", code.as_str()))?;
                writer.space(&mut w)?;
                writer.write_insn_args(args, &mut w)?;
            }
            Load { args, .. } => {
                w.write_all(b"load")?;
                writer.space(&mut w)?;
                writer.write_insn_args(args, &mut w)?;
            }
            Store { args } => {
                w.write_all(b"store")?;
                writer.write_insn_args(args, &mut w)?;
            }
            Jump { code, dest } => {
                w.write_fmt(format_args!("{}", code.as_str()))?;
                writer.space(&mut w)?;
                dest.write(writer, &mut w)?;
            }
            Branch { code, args, dest } => {
                w.write_fmt(format_args!("{}", code.as_str()))?;
                writer.space(&mut w)?;
                writer.write_insn_args(args, &mut w)?;
                writer.space(&mut w)?;
                dest.write(writer, &mut w)?;
            }
            Return { args } => {
                w.write_all(b"return")?;
                if !args.is_empty() {
                    writer.space(&mut w)?;
                }
                writer.write_insn_args(args, &mut w)?;
            }
            Phi { values, blocks, .. } => {
                w.write_all(b"phi")?;
                writer.space(&mut w)?;
                let mut args = vec![];
                for (value, block) in values.iter().zip(blocks.iter()) {
                    let mut arg = Vec::new();
                    arg.push(b'(');
                    value.write(writer, &mut arg)?;
                    writer.space(&mut arg)?;
                    block.write(writer, &mut arg)?;
                    arg.push(b')');
                    args.push(arg);
                }

                writer.write_iter_with_delim(args.iter(), " ", &mut w)?;
            }
        }

        w.write_all(b";")?;

        Ok(())
    }
}

impl<T> IrWrite for &T
where
    T: IrWrite,
{
    fn write(&self, f: &mut FuncWriter, w: impl io::Write) -> io::Result<()> {
        (*self).write(f, w)
    }
}

impl IrWrite for Vec<u8> {
    fn write(&self, _: &mut FuncWriter, mut w: impl io::Write) -> io::Result<()> {
        w.write_all(&self)
    }
}
