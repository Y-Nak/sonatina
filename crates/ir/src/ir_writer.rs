use std::io;

use super::{BlockId, Function};
use crate::{
    module::{FuncRef, ModuleCtx},
    DataFlowGraph, InstId, Module, Value, ValueId,
};

pub struct ModuleWriter<'a> {
    pub module: &'a Module,
    pub dbg: &'a dyn DebugProvider,
}

impl<'a> ModuleWriter<'a> {
    pub fn new(module: &'a Module) -> Self {
        Self::with_debug_provider(module, &DEFAULT_PROVIDER)
    }

    pub fn with_debug_provider(module: &'a Module, dbg: &'a dyn DebugProvider) -> Self {
        Self { module, dbg }
    }

    pub fn write(&mut self, w: &mut impl io::Write) -> io::Result<()> {
        // Write target.
        writeln!(w, "target = {}", self.module.ctx.triple)?;

        // Write struct types defined in the module.
        self.module.ctx.with_ty_store(|s| {
            for s in s.all_struct_data() {
                s.write(&self.module.ctx, &mut *w)?;
            }
            io::Result::Ok(())
        })?;

        // Write module level global variables.
        self.module.ctx.with_gv_store(|s| {
            for gv in s.all_gv_data() {
                gv.write(&self.module.ctx, &mut *w)?;
            }

            io::Result::Ok(())
        })?;

        for func_ref in self.module.funcs.keys() {
            let func = &self.module.funcs[func_ref];
            let mut writer = FuncWriter::with_debug_provider(func, func_ref, self.dbg);
            writer.write(&mut *w)?;
        }

        Ok(())
    }

    pub fn dump_string(&mut self) -> String {
        let mut s = Vec::new();
        self.write(&mut s).unwrap();
        unsafe { String::from_utf8_unchecked(s) }
    }
}

pub struct FuncWriteCtx<'a> {
    pub func: &'a Function,
    pub func_ref: FuncRef,
    pub dbg: &'a dyn DebugProvider,
}

impl<'a> FuncWriteCtx<'a> {
    pub fn with_debug_provider(
        func: &'a Function,
        func_ref: FuncRef,
        dbg: &'a dyn DebugProvider,
    ) -> Self {
        Self {
            func,
            func_ref,
            dbg,
        }
    }

    pub fn new(func: &'a Function, func_ref: FuncRef) -> Self {
        Self::with_debug_provider(func, func_ref, &DEFAULT_PROVIDER)
    }

    pub fn write_value(&self, value: ValueId, w: &mut impl io::Write) -> io::Result<()> {
        if let Some(name) = self.dbg.value_name(self.func, self.func_ref, value) {
            write!(w, "{name}")
        } else {
            match self.func.dfg.value(value) {
                Value::Immediate { imm, ty } => {
                    write!(w, "{}.", imm)?;
                    ty.write(self.func.ctx(), w)
                }
                Value::Global { gv, .. } => self
                    .func
                    .dfg
                    .ctx
                    .with_gv_store(|s| write!(w, "%{}", s.gv_data(*gv).symbol)),
                _ => {
                    write!(w, "v{}", value.0)
                }
            }
        }
    }

    pub fn dump_value_string(&self, value: ValueId) -> String {
        let mut s = Vec::new();
        self.write_value(value, &mut s).unwrap();
        unsafe { String::from_utf8_unchecked(s) }
    }

    pub fn dfg(&self) -> &DataFlowGraph {
        &self.func.dfg
    }

    pub fn module_ctx(&self) -> &ModuleCtx {
        self.func.ctx()
    }
}

pub struct FuncWriter<'a> {
    ctx: FuncWriteCtx<'a>,
    level: u8,
}

impl<'a> FuncWriter<'a> {
    pub fn new(func: &'a Function, func_ref: FuncRef) -> Self {
        Self::with_debug_provider(func, func_ref, &DEFAULT_PROVIDER)
    }

    pub fn with_debug_provider(
        func: &'a Function,
        func_ref: FuncRef,
        dbg: &'a dyn DebugProvider,
    ) -> Self {
        let ctx = FuncWriteCtx::with_debug_provider(func, func_ref, dbg);
        Self { ctx, level: 0 }
    }

    pub fn write(&mut self, w: &mut impl io::Write) -> io::Result<()> {
        w.write_fmt(format_args!(
            "func {} %{}(",
            self.ctx.func.sig.linkage(),
            self.ctx.func.sig.name()
        ))?;
        write_iter_with_delim(
            &mut *w,
            &self.ctx,
            self.ctx.func.arg_values.iter().map(|v| ValueWithTy(*v)),
            ", ",
        )?;
        write!(w, ") -> ")?;
        self.ctx
            .func
            .sig
            .ret_ty()
            .write(self.ctx.module_ctx(), &mut *w)?;
        writeln!(w, " {{")?;

        self.level += 1;
        for block in self.ctx.func.layout.iter_block() {
            self.write_block_with_inst(block, &mut *w)?;
        }

        self.level -= 1;
        writeln!(w, "}}")
    }

    pub fn dump_string(&mut self) -> String {
        let mut s = Vec::new();
        self.write(&mut s).unwrap();
        unsafe { String::from_utf8_unchecked(s) }
    }

    fn write_block_with_inst(&mut self, block: BlockId, w: &mut impl io::Write) -> io::Result<()> {
        self.indent(&mut *w)?;
        block.write(&self.ctx, &mut *w)?;

        self.enter(&mut *w)?;
        for inst in self.ctx.func.layout.iter_inst(block) {
            self.write_inst_in_block(inst, &mut *w)?;
        }
        self.leave();

        self.newline(w)
    }

    fn write_inst_in_block(&mut self, inst: InstId, w: &mut impl io::Write) -> io::Result<()> {
        self.indent(&mut *w)?;
        let inst_with_res = InstStatement(inst);
        inst_with_res.write(&self.ctx, &mut *w)?;
        writeln!(w)
    }

    fn indent(&self, mut w: impl io::Write) -> io::Result<()> {
        write!(w, "{}", " ".repeat(self.level as usize * 4))
    }

    fn newline(&self, mut w: impl io::Write) -> io::Result<()> {
        writeln!(w)
    }

    fn enter(&mut self, mut w: impl io::Write) -> io::Result<()> {
        self.level += 1;
        writeln!(w, ":")
    }

    fn leave(&mut self) {
        self.level -= 1;
    }
}

pub trait WriteWithModule {
    fn write(&self, module: &ModuleCtx, w: &mut impl io::Write) -> io::Result<()>;

    fn dump_string(&self, module: &ModuleCtx) -> String {
        let mut s = Vec::new();
        self.write(module, &mut s).unwrap();
        unsafe { String::from_utf8_unchecked(s) }
    }
}

impl<T> WriteWithModule for &T
where
    T: WriteWithModule,
{
    fn write(&self, module: &ModuleCtx, w: &mut impl io::Write) -> io::Result<()> {
        (*self).write(module, w)
    }
}

pub trait WriteWithFunc {
    fn write(&self, ctx: &FuncWriteCtx, w: &mut impl io::Write) -> io::Result<()>;

    fn dump_string(&self, ctx: &FuncWriteCtx) -> String {
        let mut s = Vec::new();
        self.write(ctx, &mut s).unwrap();
        unsafe { String::from_utf8_unchecked(s) }
    }
}

impl<T> WriteWithFunc for &T
where
    T: WriteWithFunc,
{
    fn write(&self, ctx: &FuncWriteCtx, w: &mut impl io::Write) -> io::Result<()> {
        (*self).write(ctx, w)
    }
}

#[derive(Clone, Copy)]
pub struct ValueWithTy(pub ValueId);
impl WriteWithFunc for ValueWithTy {
    fn write(&self, ctx: &FuncWriteCtx, w: &mut impl io::Write) -> io::Result<()> {
        let value = self.0;
        ctx.write_value(value, w)?;
        if let Value::Immediate { .. } = ctx.dfg().value(self.0) {
            Ok(())
        } else {
            let ty = ctx.dfg().value_ty(self.0);
            write!(w, ".")?;
            ty.write(ctx.module_ctx(), &mut *w)
        }
    }
}

pub struct InstStatement(pub InstId);
impl WriteWithFunc for InstStatement {
    fn write(&self, ctx: &FuncWriteCtx, w: &mut impl io::Write) -> io::Result<()> {
        if let Some(result) = ctx.func.dfg.inst_result(self.0) {
            let result_with_ty = ValueWithTy(result);
            result_with_ty.write(ctx, &mut *w)?;
            write!(w, " = ")?;
        };

        self.0.write(ctx, w)?;
        write!(w, ";")
    }
}

pub fn write_iter_with_delim<T>(
    w: &mut impl io::Write,
    ctx: &FuncWriteCtx,
    iter: impl Iterator<Item = T>,
    delim: &str,
) -> io::Result<()>
where
    T: WriteWithFunc,
{
    let mut iter = iter.peekable();
    while let Some(item) = iter.next() {
        item.write(ctx, &mut *w)?;
        if iter.peek().is_some() {
            write!(w, "{delim}")?;
        }
    }

    Ok(())
}

pub trait DebugProvider {
    #[allow(unused)]
    fn value_name(&self, func: &Function, func_ref: FuncRef, value: ValueId) -> Option<&str> {
        None
    }
}

const DEFAULT_PROVIDER: DefaultDebugProvider = DefaultDebugProvider {};

struct DefaultDebugProvider {}
impl DebugProvider for DefaultDebugProvider {}
