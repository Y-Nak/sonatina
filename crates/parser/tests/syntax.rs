use dir_test::{dir_test, Fixture};
use indenter::indented;
use ir::ir_writer::ModuleWriter;
use pest::{iterators::Pairs, Parser as _};
use sonatina_parser::{
    ast, parse_module,
    syntax::{Parser, Rule},
    Error,
};
use std::fmt::{self, Write};
mod common;

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/syntax/stmts",
    glob: "*.sntn"
)]
fn test_stmts(fixture: Fixture<&str>) {
    test_rule(Rule::_stmts, fixture)
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/syntax/func",
    glob: "*.sntn"
)]
fn test_func(fixture: Fixture<&str>) {
    test_rule(Rule::_functions, fixture)
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/syntax/module",
    glob: "*.sntn"
)]
fn test_module(fixture: Fixture<&str>) {
    test_rule(Rule::module, fixture)
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/syntax/module",
    glob: "*.sntn"
)]
fn test_module_ast(fixture: Fixture<&str>) {
    let module = ast::parse(fixture.content()).unwrap();
    snap_test!(format!("{:#?}", module), fixture.path(), Some("ast"));
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/syntax/module",
    glob: "*.sntn"
)]
fn test_module_ir(fixture: Fixture<&str>) {
    let module = parse_module(fixture.content()).unwrap();
    let mut w = ModuleWriter::with_debug_provider(&module.module, &module.debug);
    snap_test!(w.dump_string().unwrap(), fixture.path(), Some("ir"));
}

fn test_rule(rule: Rule, fixture: Fixture<&str>) {
    match Parser::parse(rule, fixture.content()) {
        Ok(r) => {
            let s = format!("{:?}", PairsWrapper(r));
            snap_test!(s, fixture.path());
        }
        Err(err) => {
            report_error(err, &fixture);
            panic!();
        }
    }
}

fn report_error(err: pest::error::Error<Rule>, fixture: &Fixture<&str>) {
    let s = Error::SyntaxError(err).print_to_string(fixture.path(), fixture.content(), true);
    eprintln!("{s}");
}

struct PairsWrapper<'i>(Pairs<'i, Rule>);

impl<'i> fmt::Debug for PairsWrapper<'i> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for pair in self.0.clone() {
            let rule = pair.as_rule();
            writeln!(f, "{rule:?} \"{}\"", pair.as_str(),)?;
            let pairs = pair.into_inner();
            if pairs.len() > 0 {
                write!(indented(f).with_str("  "), "{:?}", &PairsWrapper(pairs))?;
            }
        }
        Ok(())
    }
}
