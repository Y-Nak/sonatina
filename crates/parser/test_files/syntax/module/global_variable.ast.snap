---
source: crates/parser/tests/syntax.rs
expression: "Module {\n    target: Some(\n        TargetTriple {\n            architecture: Evm,\n            vendor: Ethereum,\n            operating_system: Evm(\n                Cancun,\n            ),\n        },\n    ),\n    declared_functions: [],\n    declared_gvs: [\n        GlobalVariable {\n            name: GvName {\n                name: \"ZERO\",\n                ..\n            },\n            ty: Type {\n                kind: Int(\n                    I256,\n                ),\n                ..\n            },\n            linkage: Private,\n            is_const: true,\n            init: Some(\n                GvInitializer {\n                    kind: Immediate(\n                        I256(\n                            I256 {\n                                is_negative: false,\n                                abs: 0,\n                            },\n                        ),\n                    ),\n                    ..\n                },\n            ),\n        },\n        GlobalVariable {\n            name: GvName {\n                name: \"PTR\",\n                ..\n            },\n            ty: Type {\n                kind: Ptr(\n                    Type {\n                        kind: Int(\n                            I256,\n                        ),\n                        ..\n                    },\n                ),\n                ..\n            },\n            linkage: Public,\n            is_const: false,\n            init: Some(\n                GvInitializer {\n                    kind: Immediate(\n                        I256(\n                            I256 {\n                                is_negative: false,\n                                abs: 1,\n                            },\n                        ),\n                    ),\n                    ..\n                },\n            ),\n        },\n        GlobalVariable {\n            name: GvName {\n                name: \"ARRAY\",\n                ..\n            },\n            ty: Type {\n                kind: Array(\n                    Type {\n                        kind: Int(\n                            I8,\n                        ),\n                        ..\n                    },\n                    4,\n                ),\n                ..\n            },\n            linkage: Private,\n            is_const: false,\n            init: Some(\n                GvInitializer {\n                    kind: Array(\n                        [\n                            GvInitializer {\n                                kind: Immediate(\n                                    I256(\n                                        I256 {\n                                            is_negative: false,\n                                            abs: 0,\n                                        },\n                                    ),\n                                ),\n                                ..\n                            },\n                            GvInitializer {\n                                kind: Immediate(\n                                    I256(\n                                        I256 {\n                                            is_negative: false,\n                                            abs: 1,\n                                        },\n                                    ),\n                                ),\n                                ..\n                            },\n                            GvInitializer {\n                                kind: Immediate(\n                                    I256(\n                                        I256 {\n                                            is_negative: false,\n                                            abs: 2,\n                                        },\n                                    ),\n                                ),\n                                ..\n                            },\n                            GvInitializer {\n                                kind: Immediate(\n                                    I256(\n                                        I256 {\n                                            is_negative: false,\n                                            abs: 3,\n                                        },\n                                    ),\n                                ),\n                                ..\n                            },\n                        ],\n                    ),\n                    ..\n                },\n            ),\n        },\n        GlobalVariable {\n            name: GvName {\n                name: \"FOO\",\n                ..\n            },\n            ty: Type {\n                kind: Struct(\n                    \"foo\",\n                ),\n                ..\n            },\n            linkage: Private,\n            is_const: false,\n            init: Some(\n                GvInitializer {\n                    kind: Struct(\n                        [\n                            GvInitializer {\n                                kind: Immediate(\n                                    I256(\n                                        I256 {\n                                            is_negative: false,\n                                            abs: 1,\n                                        },\n                                    ),\n                                ),\n                                ..\n                            },\n                            GvInitializer {\n                                kind: Immediate(\n                                    I256(\n                                        I256 {\n                                            is_negative: false,\n                                            abs: 2,\n                                        },\n                                    ),\n                                ),\n                                ..\n                            },\n                            GvInitializer {\n                                kind: Immediate(\n                                    I256(\n                                        I256 {\n                                            is_negative: false,\n                                            abs: 3,\n                                        },\n                                    ),\n                                ),\n                                ..\n                            },\n                        ],\n                    ),\n                    ..\n                },\n            ),\n        },\n        GlobalVariable {\n            name: GvName {\n                name: \"FOO_ARRAY\",\n                ..\n            },\n            ty: Type {\n                kind: Array(\n                    Type {\n                        kind: Struct(\n                            \"foo\",\n                        ),\n                        ..\n                    },\n                    2,\n                ),\n                ..\n            },\n            linkage: Private,\n            is_const: false,\n            init: Some(\n                GvInitializer {\n                    kind: Array(\n                        [\n                            GvInitializer {\n                                kind: Struct(\n                                    [\n                                        GvInitializer {\n                                            kind: Immediate(\n                                                I256(\n                                                    I256 {\n                                                        is_negative: false,\n                                                        abs: 1,\n                                                    },\n                                                ),\n                                            ),\n                                            ..\n                                        },\n                                        GvInitializer {\n                                            kind: Immediate(\n                                                I256(\n                                                    I256 {\n                                                        is_negative: false,\n                                                        abs: 2,\n                                                    },\n                                                ),\n                                            ),\n                                            ..\n                                        },\n                                        GvInitializer {\n                                            kind: Immediate(\n                                                I256(\n                                                    I256 {\n                                                        is_negative: false,\n                                                        abs: 3,\n                                                    },\n                                                ),\n                                            ),\n                                            ..\n                                        },\n                                    ],\n                                ),\n                                ..\n                            },\n                            GvInitializer {\n                                kind: Struct(\n                                    [\n                                        GvInitializer {\n                                            kind: Immediate(\n                                                I256(\n                                                    I256 {\n                                                        is_negative: false,\n                                                        abs: 3,\n                                                    },\n                                                ),\n                                            ),\n                                            ..\n                                        },\n                                        GvInitializer {\n                                            kind: Immediate(\n                                                I256(\n                                                    I256 {\n                                                        is_negative: false,\n                                                        abs: 4,\n                                                    },\n                                                ),\n                                            ),\n                                            ..\n                                        },\n                                        GvInitializer {\n                                            kind: Immediate(\n                                                I256(\n                                                    I256 {\n                                                        is_negative: false,\n                                                        abs: 5,\n                                                    },\n                                                ),\n                                            ),\n                                            ..\n                                        },\n                                    ],\n                                ),\n                                ..\n                            },\n                        ],\n                    ),\n                    ..\n                },\n            ),\n        },\n    ],\n    struct_types: [\n        Struct {\n            name: StructName(\n                \"foo\",\n            ),\n            fields: [\n                Type {\n                    kind: Int(\n                        I8,\n                    ),\n                    ..\n                },\n                Type {\n                    kind: Int(\n                        I16,\n                    ),\n                    ..\n                },\n                Type {\n                    kind: Ptr(\n                        Type {\n                            kind: Int(\n                                I64,\n                            ),\n                            ..\n                        },\n                    ),\n                    ..\n                },\n            ],\n            packed: false,\n        },\n    ],\n    functions: [],\n    comments: [],\n}"
input_file: test_files/syntax/module/global_variable.sntn
---
Module {
    target: Some(
        TargetTriple {
            architecture: Evm,
            vendor: Ethereum,
            operating_system: Evm(
                Cancun,
            ),
        },
    ),
    declared_functions: [],
    declared_gvs: [
        GlobalVariable {
            name: GvName {
                name: "ZERO",
                ..
            },
            ty: Type {
                kind: Int(
                    I256,
                ),
                ..
            },
            linkage: Private,
            is_const: true,
            init: Some(
                GvInitializer {
                    kind: Immediate(
                        I256(
                            I256 {
                                is_negative: false,
                                abs: 0,
                            },
                        ),
                    ),
                    ..
                },
            ),
        },
        GlobalVariable {
            name: GvName {
                name: "PTR",
                ..
            },
            ty: Type {
                kind: Ptr(
                    Type {
                        kind: Int(
                            I256,
                        ),
                        ..
                    },
                ),
                ..
            },
            linkage: Public,
            is_const: false,
            init: Some(
                GvInitializer {
                    kind: Immediate(
                        I256(
                            I256 {
                                is_negative: false,
                                abs: 1,
                            },
                        ),
                    ),
                    ..
                },
            ),
        },
        GlobalVariable {
            name: GvName {
                name: "ARRAY",
                ..
            },
            ty: Type {
                kind: Array(
                    Type {
                        kind: Int(
                            I8,
                        ),
                        ..
                    },
                    4,
                ),
                ..
            },
            linkage: Private,
            is_const: false,
            init: Some(
                GvInitializer {
                    kind: Array(
                        [
                            GvInitializer {
                                kind: Immediate(
                                    I256(
                                        I256 {
                                            is_negative: false,
                                            abs: 0,
                                        },
                                    ),
                                ),
                                ..
                            },
                            GvInitializer {
                                kind: Immediate(
                                    I256(
                                        I256 {
                                            is_negative: false,
                                            abs: 1,
                                        },
                                    ),
                                ),
                                ..
                            },
                            GvInitializer {
                                kind: Immediate(
                                    I256(
                                        I256 {
                                            is_negative: false,
                                            abs: 2,
                                        },
                                    ),
                                ),
                                ..
                            },
                            GvInitializer {
                                kind: Immediate(
                                    I256(
                                        I256 {
                                            is_negative: false,
                                            abs: 3,
                                        },
                                    ),
                                ),
                                ..
                            },
                        ],
                    ),
                    ..
                },
            ),
        },
        GlobalVariable {
            name: GvName {
                name: "FOO",
                ..
            },
            ty: Type {
                kind: Struct(
                    "foo",
                ),
                ..
            },
            linkage: Private,
            is_const: false,
            init: Some(
                GvInitializer {
                    kind: Struct(
                        [
                            GvInitializer {
                                kind: Immediate(
                                    I256(
                                        I256 {
                                            is_negative: false,
                                            abs: 1,
                                        },
                                    ),
                                ),
                                ..
                            },
                            GvInitializer {
                                kind: Immediate(
                                    I256(
                                        I256 {
                                            is_negative: false,
                                            abs: 2,
                                        },
                                    ),
                                ),
                                ..
                            },
                            GvInitializer {
                                kind: Immediate(
                                    I256(
                                        I256 {
                                            is_negative: false,
                                            abs: 3,
                                        },
                                    ),
                                ),
                                ..
                            },
                        ],
                    ),
                    ..
                },
            ),
        },
        GlobalVariable {
            name: GvName {
                name: "FOO_ARRAY",
                ..
            },
            ty: Type {
                kind: Array(
                    Type {
                        kind: Struct(
                            "foo",
                        ),
                        ..
                    },
                    2,
                ),
                ..
            },
            linkage: Private,
            is_const: false,
            init: Some(
                GvInitializer {
                    kind: Array(
                        [
                            GvInitializer {
                                kind: Struct(
                                    [
                                        GvInitializer {
                                            kind: Immediate(
                                                I256(
                                                    I256 {
                                                        is_negative: false,
                                                        abs: 1,
                                                    },
                                                ),
                                            ),
                                            ..
                                        },
                                        GvInitializer {
                                            kind: Immediate(
                                                I256(
                                                    I256 {
                                                        is_negative: false,
                                                        abs: 2,
                                                    },
                                                ),
                                            ),
                                            ..
                                        },
                                        GvInitializer {
                                            kind: Immediate(
                                                I256(
                                                    I256 {
                                                        is_negative: false,
                                                        abs: 3,
                                                    },
                                                ),
                                            ),
                                            ..
                                        },
                                    ],
                                ),
                                ..
                            },
                            GvInitializer {
                                kind: Struct(
                                    [
                                        GvInitializer {
                                            kind: Immediate(
                                                I256(
                                                    I256 {
                                                        is_negative: false,
                                                        abs: 3,
                                                    },
                                                ),
                                            ),
                                            ..
                                        },
                                        GvInitializer {
                                            kind: Immediate(
                                                I256(
                                                    I256 {
                                                        is_negative: false,
                                                        abs: 4,
                                                    },
                                                ),
                                            ),
                                            ..
                                        },
                                        GvInitializer {
                                            kind: Immediate(
                                                I256(
                                                    I256 {
                                                        is_negative: false,
                                                        abs: 5,
                                                    },
                                                ),
                                            ),
                                            ..
                                        },
                                    ],
                                ),
                                ..
                            },
                        ],
                    ),
                    ..
                },
            ),
        },
    ],
    struct_types: [
        Struct {
            name: StructName(
                "foo",
            ),
            fields: [
                Type {
                    kind: Int(
                        I8,
                    ),
                    ..
                },
                Type {
                    kind: Int(
                        I16,
                    ),
                    ..
                },
                Type {
                    kind: Ptr(
                        Type {
                            kind: Int(
                                I64,
                            ),
                            ..
                        },
                    ),
                    ..
                },
            ],
            packed: false,
        },
    ],
    functions: [],
    comments: [],
}
